TxDownloadManager followups

glozow opened the pull request:
Addresses some remaining followups from #30110:
- https://github.com/bitcoin/bitcoin/pull/30110#discussion_r1818893833
- https://github.com/bitcoin/bitcoin/pull/30110#discussion_r1819638959
- https://github.com/bitcoin/bitcoin/pull/30110#discussion_r1819634235

glozow committed:
[doc] comment fixups from n30110

glozow committed:
[fuzz] allow negative time jumps in txdownloadman_impl

instagibbs commented:
is this ready for undraft?

naumenkogs reviewed (commented):
ACK 8351562bec6081eda2a600bfe4edeb264a9dee0b

I haven't verified what happens if `time` goes below 0

dergoegge commented:
oss-fuzz found a crash in txdownloadman_impl, perhaps worth including the fix here or in another followup:

```
$ FUZZ=txdownloadman_impl out/libfuzzer/fuzz clusterfuzz-testcase-minimized-txdownloadman_impl-5180010463297536.txt
src/test/fuzz/txdownloadman.cpp:290 CheckInvariants: Assertion `txdownload_impl.m_txrequest.CountInFlight(peer) <= node::MAX_PEER_TX_REQUEST_IN_FLIGHT' failed.
```

[clusterfuzz-testcase-minimized-txdownloadman_impl-5180010463297536.txt](https://github.com/user-attachments/files/17645609/clusterfuzz-testcase-minimized-txdownloadman_impl-5180010463297536.txt)

dergoegge commented:
> is that on master

Sorry I should have clarified, it is on master.

glozow commented:
Thanks @dergoegge, I probably also could have inferred from it being oss-fuzz.

Even though I wrote it, the assertion looks wrong 😅 overloaded inflight is handled with just a delay

https://github.com/bitcoin/bitcoin/blob/80cb630bd945e7f1777c7bab041c411b6916e651/src/node/txdownloadman_impl.cpp#L199

I think what I was trying to do was ensure `MAX_PEER_TX_ANNOUNCEMENTS` isn't busted. Adding a fix here.

glozow committed:
fuzz fix: assert MAX_PEER_TX_ANNOUNCEMENTS is not exceeded

Previously this assertion checked MAX_PEER_TX_REQUEST_IN_FLIGHT was not
exceeded. However, this property is not actually enforced; it is just
used to determine when a peer is overloaded.

instagibbs reviewed (approved):
ACK 5dc94d13d419b8d5e543cb50edeb872335c090e7

instagibbs commented on src/test/fuzz/txdownloadman.cpp:
```diff
@@ -287,7 +287,7 @@ static void CheckInvariants(const node::TxDownloadManagerImpl& txdownload_impl,
     for (NodeId peer = 0; peer < NUM_PEERS; ++peer) {
         if (!HasRelayPermissions(peer)) {
-            Assert(txdownload_impl.m_txrequest.CountInFlight(peer) <= node::MAX_PEER_TX_REQUEST_IN_FLIGHT);
+            Assert(txdownload_impl.m_txrequest.Count(peer) <= node::MAX_PEER_TX_ANNOUNCEMENTS);
```
any value in 
```suggestion
            Assert(txdownload_impl.m_txrequest.CountInFlight(peer) <= node::MAX_PEER_TX_ANNOUNCEMENTS);
            Assert(txdownload_impl.m_txrequest.Count(peer) <= node::MAX_PEER_TX_ANNOUNCEMENTS);
```
?

glozow replied:
What do you mean any value in?

instagibbs replied:
would the suggested change potentially give coverage

naumenkogs replied:
Perhaps `CountInFlight` should check against `MAX_PEER_TX_REQUEST_IN_FLIGHT` and `Count` against `MAX_PEER_TX_ANNOUNCEMENTS`, separately?
I think that's what Greg wanted to suggest.

marcofleon replied:
Could be misunderstanding here, but wouldn't `m_txrequest.Count(peer)` always be greater than (or equal?) to `m_txrequest.CountInFlight(peer)`, so the additional assert would be redundant?

theStack reviewed (approved):
ACK 5dc94d13d419b8d5e543cb50edeb872335c090e7

Joj501 reviewed (approved):
Connect with my wallet address

naumenkogs commented:
ACK https://github.com/bitcoin/bitcoin/commit/5dc94d13d419b8d5e543cb50edeb872335c090e7

marcofleon commented on src/test/fuzz/txdownloadman.cpp:
```diff
@@ -287,7 +287,7 @@ static void CheckInvariants(const node::TxDownloadManagerImpl& txdownload_impl,
     // We should never have more than the maximum in-flight requests out for a peer.
```
Update comment for the new assert?

instagibbs commented:
Assuming we implemented everything correctly yes!

On Thu, Nov 7, 2024, 6:42 AM marcofleon ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/test/fuzz/txdownloadman.cpp
> <https://github.com/bitcoin/bitcoin/pull/31190#discussion_r1832534828>:
>
> > @@ -287,7 +287,7 @@ static void CheckInvariants(const node::TxDownloadManagerImpl& txdownload_impl,
>      // We should never have more than the maximum in-flight requests out for a peer.
>      for (NodeId peer = 0; peer < NUM_PEERS; ++peer) {
>          if (!HasRelayPermissions(peer)) {
> -            Assert(txdownload_impl.m_txrequest.CountInFlight(peer) <= node::MAX_PEER_TX_REQUEST_IN_FLIGHT);
> +            Assert(txdownload_impl.m_txrequest.Count(peer) <= node::MAX_PEER_TX_ANNOUNCEMENTS);
>
> Could be misunderstanding here, but wouldn't m_txrequest.Count(peer)
> always be greater than (or equal?) to m_txrequest.CountInFlight(peer), so
> the additional assert would be redundant?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/31190#discussion_r1832534828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMAFU7JC45F4HYWMEIFEK3Z7NGZXAVCNFSM6AAAAABQ6GY2POVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDIMRQG4ZTEMBWGA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

marcofleon commented:
> Assuming we implemented everything correctly yes!

hmm... you make a good point
